File: backend/Booking/src/Modules/Booking.Modules.Catalog/Features/CatalogEndpoints.cs
namespace Booking.Modules.Catalog.Features;

public static class CatalogEndpoints
{
    private const string Base = "api";

    public static class Stores
    {
        public const string Create = Base + "/stores";
        public const string Update = Base + "/stores";
        public const string CheckSlugAvailability = Base + "/stores/slug-availability/{slug}";
        public const string UpdatePicture = Base + "/stores/picture";
        public const string GetMy = Base + "/stores/me";

        public const string GetPublic = Base + "/stores/{slug}";
    }

    public static class Products
    {
        public static class Sessions
        {
            public const string Create = $"{Base}/products/s/create";
            public const string Update = $"{Base}/products/s/{{productSlug}}";
            public const string GetMy = $"{Base}/products/s/{{productSlug}}/private";
            public const string GetAllSessions = $"{Base}/products/s"; // get booked session for the store owner 
            // queryParam :string? month ,string?year , string? timeZoneId


            // PUBLIC : 
            public const string Get = $"{Base}/products/s/{{productSlug}}";
            public const string Book = $"{Base}/products/s/{{productSlug}}";
            public const string GetDailyAv = $"{Base}/products/s/{{productSlug}}/availability"; // Query: date

            public const string
                GetMonthlyAv =
                    $"{Base}/products/s/{{productSlug}}/availability/month";
        }

        public const string Arrange = $"{Base}/produtcs/arrange"; 
        public const string Delete = $"{Base}/products/{{productSlug}}";
        public const string TogglePublished = $"{Base}/products/toggle/{{productSlug}}";
    }

    public static class Orders
    {
        public const string Get = $"{Base}" + "/orders/{orderId:int}";
        public const string GetMy = $"{Base}" + "/orders/me";
    }

    public static class Payouts
    {
        public const string Payout = $"{Base}/payout";
        public const string PayoutHistory = $"{Base}/payout";

        public static class Admin
        {
            public const string ApprovePayout = $"{Base}/admin/payout/approve"; // body : payoutId
            public const string WebhookPayout = $"{Base}/admin/payout/webhook";
            public const string RejectPayout = $"{Base}/admin/payout/reject"; // body : payoutId 

            public const string
                GetAllPayouts =
                    $"{Base}/admin/payout"; // query : status (Pending, Approved, Rejected , Completed) , upToDate , timeZoneId
            // public const string GetDetailled = $"{Base}/admin/payout/{{payoutId}}";
        }
    }

    public static class Payment
    {
        public const string Create = $"{Base}/payments";
        public const string GetWallet = $"{Base}/payments/wallet";
        public const string Webhook = $"{Base}/payments/webhook"; //  payment_ref=5f9498735289e405fc7c18ac
    }
}

File: frontend/src/api/utils/catalog-endpoints.ts
/**
 * Catalog API Endpoints
 * Based on backend CatalogEndpoints.cs
 */

const BASE = 'api';

export const CatalogEndpoints = {
  Stores: {
    Create: `${BASE}/stores`,
    Update: `${BASE}/stores`,
    CheckSlugAvailability: (slug: string) => `${BASE}/stores/slug-availability/${slug}`,
    UpdatePicture: `${BASE}/stores/picture`,
    GetMy: `${BASE}/stores/me`,
    GetPublic: (slug: string) => `${BASE}/stores/${slug}`,
  },

  Products: {
    Sessions: {
      Create: `${BASE}/products/s/create`,
      Update: (productSlug: string) => `${BASE}/products/s/${productSlug}`,
      GetMy: (productSlug: string) => `${BASE}/products/s/${productSlug}/private`,
      GetAllSessionsMonthly: () => `${BASE}/products/s`, // queryParam :string? month ,string?year , string? timeZoneId

      // Public endpoints
      Get: (productSlug: string) => `${BASE}/products/s/${productSlug}`,
      Book: (productSlug: string) => `${BASE}/products/s/${productSlug}`,
      GetMonthlyAvailability: (productSlug: string) => `${BASE}/products/s/${productSlug}/availability/month`, // with query params year, month timeZoneId
    },

    Arrange: `${BASE}/produtcs/arrange`, // Note: typo in backend endpoint
    Delete: (productSlug: string) => `${BASE}/products/${productSlug}`,
    TogglePublished: (productSlug: string) => `${BASE}/products/toggle/${productSlug}`,
  },

  Orders: {
    Get: (orderId: number) => `${BASE}/orders/${orderId}`,
    GetMy: `${BASE}/orders/me`,
    GetOrders: `${BASE}/orders`, // query params: startsAt, endsAt
  },

  Payouts: {
    Payout: `${BASE}/payout`,
    PayoutHistory: `${BASE}/payout`,
    Admin: {
      ApprovePayout: `${BASE}/admin/payout/approve`,
      WebhookPayout: `${BASE}/admin/payout/webhook`,
      RejectPayout: `${BASE}/admin/payout/reject`,
      GetAllPayouts: `${BASE}/admin/payout`,
    },
  },

  Payment: {
    Create: `${BASE}/payments`,
    GetWallet: `${BASE}/payments/wallet`,
    Webhook: `${BASE}/payments/webhook`,
  },
} as const;

/**
 * Query parameter builders for API endpoints
 */
export const QueryBuilders = {
  Stores: {
    /**
     * Build query parameters for checking slug availability
     */
    slugAvailability: (slug: string) => ({ slug }),
  },

  Products: {
    Sessions: {
      /**
       * Build query parameters for daily availability
       */
      dailyAvailability: (date: string) => ({ date }),

      /**
       * Build query parameters for monthly availability
       */
      monthlyAvailability: (year: number, month: number, timeZoneId: string = 'Africa/Tunis') => ({
        year,
        month,
        timeZoneId,
      }),

      /**
       * Build query parameters for getting all sessions
       */
      getAllSessions: (year: number, month: number, timeZoneId: string = 'Africa/Tunis') => ({
        year,
        month,
        timeZoneId,
      }),
    },
  },

  Orders: {
    /**
     * Build query parameters for getting orders
     */
    getOrders: (startsAt?: Date, endsAt?: Date) => ({
      ...(startsAt && { startsAt: startsAt.toISOString() }),
      ...(endsAt && { endsAt: endsAt.toISOString() }),
    }),
  },

  Payouts: {
    Admin: {
      /**
       * Build query parameters for getting all payouts
       */
      getAllPayouts: (status?: 'Pending' | 'Approved' | 'Rejected' | 'Completed', upToDate?: string, timeZoneId?: string) => ({
        ...(status && { status }),
        ...(upToDate && { upToDate }),
        ...(timeZoneId && { timeZoneId }),
      }),
    },
  },
} as const;
